walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,113 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
+ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    ArrayList<Event> cleanedEvents = removeNonrequiredMeeting(events, request);
+    cleanedEvents = removeOverlap(cleanedEvents);
+
+    ArrayList<TimeRange> toReturn = new ArrayList<TimeRange>();
+    int start = TimeRange.START_OF_DAY;
+    boolean endDay = true;
+    int endTime = 0;
+    long meetingLen = request.getDuration();
+
+    Iterator<Event> iter = cleanedEvents.iterator();
+
+    if (events.isEmpty() || cleanedEvents.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    while (iter.hasNext()) {
+      Event e = iter.next();
+      
+      int eventStart = e.getWhen().start();
+      int eventEnd = eventStart + e.getWhen().duration();
+
+      //account for event at beginning of day
+      if (eventStart == start) {
+        start = eventStart + e.getWhen().duration(); 
+      } else {
+        if (eventStart - start >= meetingLen) {
+        TimeRange toAdd = TimeRange.fromStartEnd(start, eventStart, false);
+        toReturn.add(toAdd);
+        }
+        start = eventEnd;
+        
+        //check if there's available meeting time after last event
+        if (eventStart + e.getWhen().duration() < TimeRange.END_OF_DAY) {
+          endDay = false;
+          endTime = eventEnd;
+        }
+      }
+    }
+
+    //add meeting after last event
+    if (!endDay) {
+      toReturn.add(TimeRange.fromStartEnd(endTime, TimeRange.END_OF_DAY, true));
+    }
+
+    return toReturn;  
+  }
+
+  private boolean isOverlap(Event e1, Event e2) {
+    return e1.getWhen().start() + e1.getWhen().duration() > e2.getWhen().start();
+  }
+
+  private ArrayList<Event> removeNonrequiredMeeting(Collection<Event> events, MeetingRequest meeting) {
+    ArrayList<Event> newEventList = new ArrayList<>();
```

Comment:
This is ok but take a look at filter function
https://www.baeldung.com/java-collection-filtering

Give it a shot if you could!